tests/Unit/CrudPanel/CrudPanelButtonsTest.php

```diff
@@ -0,0 +1,218 @@
+    }
+
+    public function testAddButtonFromModelFunction()
+    {
+        $expectedButton = $this->topModelFunctionButton;
+
+        $this->crudPanel->addButton($expectedButton->stack, $expectedButton->name, $expectedButton->type, $expectedButton->content);
+
+        $this->assertEquals($expectedButton, $this->crudPanel->buttons->last());
+    }
+
+    public function testAddButtonFromView()
+    {
+        $expectedButton = $this->topViewButton;
+        $viewName = 'someViewName';
+
+        $this->crudPanel->addButtonFromView($expectedButton->stack, $expectedButton->name, $viewName, $expectedButton->content);
+
+        $backpackButtonViewPackage = 'vendor.backpack.crud.buttons.';
+        $actualButton = $this->crudPanel->buttons->last();
+
+        $this->assertEquals($expectedButton->stack, $actualButton->stack);
+        $this->assertEquals($expectedButton->name, $actualButton->name);
+        $this->assertEquals($backpackButtonViewPackage.$viewName, $actualButton->content);
+        $this->assertEquals($expectedButton->stack, $actualButton->stack);
+    }
+
+    public function testRemoveButton()
+    {
+        $buttonName = $this->defaultButtonNames[0];
+
+        $this->crudPanel->removeButton($buttonName);
+
+        $this->assertEquals(count($this->defaultButtonNames) - 1, count($this->crudPanel->buttons));
+        $this->assertNull($this->getButtonByName($buttonName));
+    }
+
+    public function testRemoveUnknownButton()
+    {
+        $buttonName = 'someButtonName';
+
+        $this->crudPanel->removeButton($buttonName);
+
+        $this->assertEquals(count($this->defaultButtonNames), count($this->crudPanel->buttons));
+    }
+
+    public function testRemoveAllButtons()
+    {
+        $this->crudPanel->removeAllButtons();
+
+        $this->assertEmpty($this->crudPanel->buttons);
+    }
+
+    public function testRemoveButtonFromStack()
+    {
+        $button = $this->crudPanel->buttons->first();
+
+        $this->crudPanel->removeButtonFromStack($button->name, $button->stack);
+
+        $this->assertEquals(count($this->defaultButtonNames) - 1, count($this->crudPanel->buttons));
+        $this->assertNull($this->getButtonByName($button->name));
+    }
+
+    public function testRemoveUnknownButtonFromStack()
+    {
+        $this->crudPanel->removeButtonFromStack('someButtonName', 'line');
+
+        $this->assertEquals(count($this->defaultButtonNames), count($this->crudPanel->buttons));
+    }
+
+    public function testRemoveButtonFromUnknownStack()
+    {
+        $button = $this->crudPanel->buttons->first();
+
+        $this->crudPanel->removeButtonFromStack($button->name, 'someStackName');
+
+        $this->assertEquals(count($this->defaultButtonNames), count($this->crudPanel->buttons));
+    }
+
+    public function testRemoveAllButtonsFromStack()
+    {
+        $this->crudPanel->removeAllButtonsFromStack('line');
+
+        $this->assertEquals(2, count($this->crudPanel->buttons));
```

Comment:
Where'd that 2 come from? I'd comment there as a 'remove all' implies '0' to me and this is unexpected (well it's probably expected but not intuitive).

Reply:
The method removes all buttons from a given stack. Sure, a better name would probably be `removeButtonsFromStack($stackName)`.
The crudPanel initializes by default 6 buttons, 4 on the 'line' stack and 2 on the 'top'. This removes the ones from the 'line' and checks if only 2 buttons remain.
I can do an additional check to see if the remaining buttons are indeed on the 'top' stack.

Reply:
An additional check would probably be good to do in case there are unexpected side effects.